File: javatest/omnia/data/structure/immutable/ImmutableListTest.java
package omnia.data.structure.immutable;

import omnia.data.structure.List;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

import java.util.Iterator;
import java.util.stream.Stream;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotEquals;
import static org.junit.Assert.assertSame;
import static org.junit.Assert.assertTrue;

@RunWith(JUnit4.class)
public class ImmutableListTest {

  private static final String TEST_ITEM_1 = "first";
  private static final String TEST_ITEM_2 = "second";
  private static final String TEST_ITEM_3 = "third";

  @Test
  public void getAt_whenFirst_didReturnFirst() {
    List<String> underTest = createTestSubject();

    assertSame(TEST_ITEM_1, underTest.itemAt(0));
  }

  @Test
  public void getAt_whenMiddle_didReturnMiddle() {
    List<String> underTest = createTestSubject();

    assertSame(TEST_ITEM_2, underTest.itemAt(1));
  }

  @Test
  public void getAt_whenLast_didReturnLast() {
    List<String> underTest = createTestSubject();

    assertSame(TEST_ITEM_3, underTest.itemAt(2));
  }

  @Test(expected = IndexOutOfBoundsException.class)
  public void getAt_whenEmpty_didThrowException() {
    ImmutableList.builder().build().itemAt(0);
  }

  @Test(expected = IndexOutOfBoundsException.class)
  public void getAt_negative_didThrowException() {
    createTestSubject().itemAt(-1);
  }

  @Test(expected = IndexOutOfBoundsException.class)
  public void getAt_tooHigh_didThrowException() {
    createTestSubject().itemAt(20);
  }

  @Test
  public void count_whenEmpty_didReturnCount() {
    assertEquals(0, ImmutableList.builder().build().count());
  }

  @Test
  public void count_didReturnCount() {
    assertEquals(3, createTestSubject().count());
  }

  @Test
  public void contains_whenFirst_didReturnTrue() {
    assertTrue(createTestSubject().contains(TEST_ITEM_1));
  }

  @Test
  public void contains_whenMiddle_didReturnTrue() {
    assertTrue(createTestSubject().contains(TEST_ITEM_2));
  }

  @Test
  public void contains_whenLast_didReturnTrue() {
    assertTrue(createTestSubject().contains(TEST_ITEM_3));
  }

  @Test
  public void contains_whenUnknown_didReturnFalse() {
    assertFalse(createTestSubject().contains("unknown"));
  }

  @Test
  public void iterate_didIterateOverAll() {
    Iterator<String> iterator = createTestSubject().iterator();

    assertSame(TEST_ITEM_1, iterator.next());
    assertSame(TEST_ITEM_2, iterator.next());
    assertSame(TEST_ITEM_3, iterator.next());
    assertFalse(iterator.hasNext());
  }

  @Test
  public void stream_didStreamAll() {
    Stream<String> stream = createTestSubject().stream();

    Object[] contents = stream.toArray();

    assertEquals(3, contents.length);
    assertSame(TEST_ITEM_1, contents[0]);
    assertSame(TEST_ITEM_2, contents[1]);
    assertSame(TEST_ITEM_3, contents[2]);
  }

  @Test
  public void equals_whenEmpty_didEqual() {
    assertEquals(ImmutableList.builder().build(), ImmutableList.builder().build());
  }

  @Test
  public void equals_whenEqual_areEqual() {
    assertEquals(createTestSubject(), createTestSubject());
  }

  @Test
  public void equals_whenUnequal_areNotEqual() {
    assertNotEquals(ImmutableList.<String>builder().add("hi there").build(), createTestSubject());
  }

  @Test
  public void hashCode_whenEqual_areEqual() {
    assertEquals(createTestSubject().hashCode(), createTestSubject().hashCode());
  }

  private static ImmutableList<String> createTestSubject() {
    return ImmutableList.<String>builder()
        .add(TEST_ITEM_1)
        .add(TEST_ITEM_2)
        .add(TEST_ITEM_3)
        .build();
  }
}

File: tests/omnia/data/structure/immutable/ImmutableListTest.kt
package omnia.data.structure.immutable

import com.google.common.truth.Truth.assertThat
import kotlin.test.Test
import kotlin.test.assertFailsWith
import omnia.data.structure.List

class ImmutableListTest {

  @Test
  fun getAt_whenFirst_didReturnFirst() {
    val underTest: List<String> = createTestSubject()
    assertThat(underTest.itemAt(0)).isSameInstanceAs(TEST_ITEM_1)
  }

  @Test
  fun getAt_whenMiddle_didReturnMiddle() {
    val underTest: List<String> = createTestSubject()
    assertThat(underTest.itemAt(1)).isSameInstanceAs(TEST_ITEM_2)
  }

  @Test
  fun getAt_whenLast_didReturnLast() {
    val underTest: List<String> = createTestSubject()
    assertThat(underTest.itemAt(2)).isSameInstanceAs(TEST_ITEM_3)
  }

  @Test
  fun getAt_whenEmpty_didThrowException() {
    assertFailsWith(IndexOutOfBoundsException::class) {
      ImmutableList.builder<Any>().build().itemAt(0)
    }
  }

  @Test
  fun getAt_negative_didThrowException() {
    assertFailsWith(IndexOutOfBoundsException::class) { createTestSubject().itemAt(-1) }
  }

  @Test
  fun getAt_tooHigh_didThrowException() {
    assertFailsWith(IndexOutOfBoundsException::class) { createTestSubject().itemAt(20) }
  }

  @Test
  fun count_whenEmpty_didReturnCount() {
    assertThat(ImmutableList.builder<Any>().build().count().toLong()).isEqualTo(0)
  }

  @Test
  fun count_didReturnCount() {
    assertThat(createTestSubject().count().toLong()).isEqualTo(3)
  }

  @Test
  fun contains_whenFirst_didReturnTrue() {
    assertThat(createTestSubject().contains(TEST_ITEM_1)).isTrue()
  }

  @Test
  fun contains_whenMiddle_didReturnTrue() {
    assertThat(createTestSubject().contains(TEST_ITEM_2)).isTrue()
  }

  @Test
  fun contains_whenLast_didReturnTrue() {
    assertThat(createTestSubject().contains(TEST_ITEM_3)).isTrue()
  }

  @Test
  fun contains_whenUnknown_didReturnFalse() {
    assertThat(createTestSubject().contains("unknown")).isFalse()
  }

  @Test
  fun iterate_didIterateOverAll() {
    val iterator = createTestSubject().iterator()
    assertThat(iterator.next()).isSameInstanceAs(TEST_ITEM_1)
    assertThat(iterator.next()).isSameInstanceAs(TEST_ITEM_2)
    assertThat(iterator.next()).isSameInstanceAs(TEST_ITEM_3)
    assertThat(iterator.hasNext()).isFalse()
  }

  @Test
  fun stream_didStreamAll() {
    val stream = createTestSubject().stream()
    val contents = stream.toArray()
    assertThat(contents.size.toLong()).isEqualTo(3)
    assertThat(contents[0]).isSameInstanceAs(TEST_ITEM_1)
    assertThat(contents[1]).isSameInstanceAs(TEST_ITEM_2)
    assertThat(contents[2]).isSameInstanceAs(TEST_ITEM_3)
  }

  @Test
  fun equals_whenEmpty_didEqual() {
    assertThat(ImmutableList.builder<Any>().build()).isEqualTo(ImmutableList.builder<Any>().build())
  }

  @Test
  fun equals_whenEqual_areEqual() {
    assertThat(createTestSubject()).isEqualTo(createTestSubject())
  }

  @Test
  fun equals_whenUnequal_areNotEqual() {
    assertThat(createTestSubject())
        .isNotEqualTo(ImmutableList.builder<String>().add("hi there").build())
  }

  @Test
  fun hashCode_whenEqual_areEqual() {
    assertThat(createTestSubject().hashCode().toLong())
        .isEqualTo(createTestSubject().hashCode().toLong())
  }

  companion object {

    private const val TEST_ITEM_1 = "first"
    private const val TEST_ITEM_2 = "second"
    private const val TEST_ITEM_3 = "third"
    private fun createTestSubject(): ImmutableList<String> {
      return ImmutableList.builder<String>()
        .add(TEST_ITEM_1)
        .add(TEST_ITEM_2)
        .add(TEST_ITEM_3)
        .build()
    }
  }
}